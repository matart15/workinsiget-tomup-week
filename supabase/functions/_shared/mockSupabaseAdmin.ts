const SUPABASE_FUNCTION_NAMES = [
  "from",
  "select",
  "eq",
  "delete",
  "update",
  "insert",
  "limit",
  "single",
] as const;
type ReturnFunctionName = (typeof SUPABASE_FUNCTION_NAMES)[number];

export const mockSupabaseAdmin: any = (
  returnValues: Record<ReturnFunctionName, any>[],
) => {
  let allCalls: any = [];
  let allArgs: any = [];
  let currentChainIndex = 0;

  const supabaseAdmin: any = {};
  for (const functionName of SUPABASE_FUNCTION_NAMES) {
    supabaseAdmin[functionName] = (...args: any[]) => {
      const returnValue = returnValues[currentChainIndex];
      const returnFunctionName: ReturnFunctionName = Object.keys(
        returnValue,
      )[0] as any;

      allArgs.push({ [functionName]: args });
      if (returnFunctionName === functionName) {
        allCalls.push(allArgs);
        currentChainIndex++;
        return returnValue[returnFunctionName];
      }
      return supabaseAdmin;
    };
  }

  return {
    allCalls,
    supabaseAdmin,
  };
};
