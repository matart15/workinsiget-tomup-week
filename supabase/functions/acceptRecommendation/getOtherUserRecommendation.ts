import { SupabaseClient } from "https://esm.sh/@supabase/supabase-js@2";

import { Database } from "../_shared/database.ts";

export async function getOtherUserRecommendation({
  supabaseAdmin,
  otherUserId,
  currentUserId,
}: {
  supabaseAdmin: SupabaseClient<Database>;
  otherUserId: string;
  currentUserId: string;
}) {
  const otherUserRecommendation = await supabaseAdmin
    .from("u_recommendations")
    .select("*")
    .eq("user1_id", otherUserId)
    .eq("user2_id", currentUserId)
    .eq("status", "ACCEPTED")
    .limit(1);

  if (otherUserRecommendation.error) {
    throw otherUserRecommendation.error;
  }

  if (otherUserRecommendation.data.length > 1) {
    throw new Error(
      "otherUserRecommendation.data.length is more than 1, this should not happen",
    );
  }
  return otherUserRecommendation;
}
