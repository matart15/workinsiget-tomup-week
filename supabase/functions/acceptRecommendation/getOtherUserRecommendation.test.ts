import { assertEquals } from "https://deno.land/std@0.206.0/testing/asserts.ts";
import { assertSnapshot } from "https://deno.land/std@0.206.0/testing/snapshot.ts";

import { mockSupabaseAdmin } from "../_shared/mockSupabaseAdmin.ts";
import { getOtherUserRecommendation } from "./getOtherUserRecommendation.ts";

Deno.test(
  "getOtherUserRecommendation should return recommendation",
  async (t) => {
    const { supabaseAdmin, allCalls } = mockSupabaseAdmin([
      {
        limit: {
          data: [{ id: 123 }],
        },
      },
    ]);
    // Act
    const result = await getOtherUserRecommendation({
      supabaseAdmin,
      otherUserId: "456",
      currentUserId: "789",
    });
    // Assert
    assertEquals(result.data, [{ id: 123 }] as any);
    // assert all args
    await assertSnapshot(t, allCalls);
  },
);
Deno.test(
  "getOtherUserRecommendation should throw an error if data length is more than 1",
  async (t) => {
    const { supabaseAdmin, allCalls } = mockSupabaseAdmin([
      {
        limit: {
          data: [{ id: "123" }, { id: "456" }],
        },
      },
    ]);

    try {
      // Act
      await getOtherUserRecommendation({
        supabaseAdmin,
        otherUserId: "456",
        currentUserId: "789",
      });
      // Assert
      throw new Error("This should not happen 1");
    } catch (error) {
      // Assert
      assertEquals(
        error.message,
        "otherUserRecommendation.data.length is more than 1, this should not happen",
      );
      // assert args
      await assertSnapshot(t, allCalls);
    }
  },
);
Deno.test(
  "getOtherUserRecommendation should throw an error if there is an error",
  async (t) => {
    const { supabaseAdmin, allCalls } = mockSupabaseAdmin([
      {
        limit: {
          error: new Error("Some error"),
        },
      },
    ]);

    try {
      // Act
      await getOtherUserRecommendation({
        supabaseAdmin,
        otherUserId: "456",
        currentUserId: "789",
      });
      // Assert
      throw new Error("This should not happen 2");
    } catch (error) {
      // Assert
      assertEquals(error.message, "Some error");
      // assert args
      await assertSnapshot(t, allCalls);
    }
  },
);
