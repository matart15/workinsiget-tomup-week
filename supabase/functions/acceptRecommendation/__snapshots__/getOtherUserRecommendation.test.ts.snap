export const snapshot = {};

snapshot[`getOtherUserRecommendation should return recommendation 1`] = `
[
  [
    {
      from: [
        "u_recommendations",
      ],
    },
    {
      select: [
        "*",
      ],
    },
    {
      eq: [
        "user1_id",
        "456",
      ],
    },
    {
      eq: [
        "user2_id",
        "789",
      ],
    },
    {
      eq: [
        "status",
        "ACCEPTED",
      ],
    },
    {
      limit: [
        1,
      ],
    },
  ],
]
`;

snapshot[`getOtherUserRecommendation should throw an error if data length is more than 1 1`] = `
[
  [
    {
      from: [
        "u_recommendations",
      ],
    },
    {
      select: [
        "*",
      ],
    },
    {
      eq: [
        "user1_id",
        "456",
      ],
    },
    {
      eq: [
        "user2_id",
        "789",
      ],
    },
    {
      eq: [
        "status",
        "ACCEPTED",
      ],
    },
    {
      limit: [
        1,
      ],
    },
  ],
]
`;

snapshot[`getOtherUserRecommendation should throw an error if there is an error 1`] = `
[
  [
    {
      from: [
        "u_recommendations",
      ],
    },
    {
      select: [
        "*",
      ],
    },
    {
      eq: [
        "user1_id",
        "456",
      ],
    },
    {
      eq: [
        "user2_id",
        "789",
      ],
    },
    {
      eq: [
        "status",
        "ACCEPTED",
      ],
    },
    {
      limit: [
        1,
      ],
    },
  ],
]
`;
